documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#sequentialDistribution--[sequentialDistribution]
+or link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#parallelDistribution--[parallelDistribution]
+methods on the `CacheStream`.   Note that currently intermediate operations return a `Stream` instance
+so you must make sure to invoke these methods before calling another intermediate operation.
+
+
+==== Task timeout
+
+It is possible to set a timeout value for the operation requests. This timeout is used only for remote requests timing out and
+it is on a per request basis. The former means the local execution will not timeout and the latter means if you have a failover
+scenario as described above the subsequent requests each have a new timeout.  If no timeout is specified, it uses the
+replication timeout as a default timeout. You can set the timeout in your task by doing the following: 
+
+[source,java]
+----
+CacheStream<Object, String> stream = cache.entrySet().stream();
+stream.timeout(1, TimeUnit.MINUTES);
+----
+
+For more information about this, please check the java doc in
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#timeout-long-java.util.concurrent.TimeUnit-[timeout]
+javadoc.
+
+==== Injection
+
+The link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html[Stream]
+has a terminal operation called
+link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-[forEach]
+which allows for running some sort of side effect operation on the data.  In this case it may be desirable to get a reference to
+the `Cache` that is backing this Stream.  In this case if your `Consumer` implements the 
+link:http://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/stream/CacheAware.html[CacheAware]
+interface the `injectCache` method be invoked before the accept method from the `Consumer` interface.
+
+There are plans to possibly adding a new forEach that instead takes a
+link:https://docs.oracle.com/javase/8/docs/api/java/util/function/BiConsumer.html[BiConsumer]
+which would allow for much cleaner API usage since a lambda could be used.
+
+==== Distributed Stream execution
+
+Distributed streams execution works in a fashion very similiar to map reduce.  Except in this case we are sending zero to many intermediate operations
+(map, filter etc.) and a single terminal operation to the various nodes.  The operation basically comes down to the following:
+
+. The desired segments are grouped by which node is the primary owner of the given segment
+
+. A request is generated to send to each remote node that contains the intermediate and terminal operations including which segments it should process
+
+.. The thread the terminal operation was invoked upon will perform the local operation directly
```

Comment:
Is this always single-threaded? Even if it is, I'd only mention nodes in this high-level description.